hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogic.java

```diff
@@ -0,0 +1,246 @@
+import static com.hedera.services.store.tokens.TokenStore.MISSING_TOKEN;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.ACCOUNT_EXPIRED_AND_PENDING_REMOVAL;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.CURRENT_TREASURY_STILL_OWNS_NFTS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_EXPIRATION_TIME;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_IS_IMMUTABLE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_IS_PAUSED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_WAS_DELETED;
+
+import com.hedera.services.context.SideEffectsTracker;
+import com.hedera.services.exceptions.InvalidTransactionException;
+import com.hedera.services.ledger.SigImpactHistorian;
+import com.hedera.services.state.enums.TokenType;
+import com.hedera.services.state.merkle.MerkleToken;
+import com.hedera.services.store.contracts.WorldLedgers;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.NftId;
+import com.hedera.services.store.tokens.HederaTokenStore;
+import com.hedera.services.store.tokens.annotations.AreTreasuryWildcardsEnabled;
+import com.hedera.services.txns.util.TokenUpdateValidator;
+import com.hedera.services.txns.validation.OptionValidator;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.hederahashgraph.api.proto.java.TokenUpdateTransactionBody;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Optional;
+import javax.inject.Inject;
+
+public class TokenUpdateLogic {
+    private final OptionValidator validator;
+    private final HederaTokenStore tokenStore;
+    private final WorldLedgers worldLedgers;
+    private final SideEffectsTracker sideEffectsTracker;
+    private final SigImpactHistorian sigImpactHistorian;
+    boolean allowChangedTreasuryToOwnNfts;
+
+    @Inject
+    public TokenUpdateLogic(
+            final @AreTreasuryWildcardsEnabled boolean allowChangedTreasuryToOwnNfts,
+            OptionValidator validator,
+            HederaTokenStore tokenStore,
+            WorldLedgers worldLedgers,
+            SideEffectsTracker sideEffectsTracker,
+            SigImpactHistorian sigImpactHistorian) {
+        this.validator = validator;
+        this.tokenStore = tokenStore;
+        this.worldLedgers = worldLedgers;
+        this.sideEffectsTracker = sideEffectsTracker;
+        this.sigImpactHistorian = sigImpactHistorian;
+        this.allowChangedTreasuryToOwnNfts = allowChangedTreasuryToOwnNfts;
+    }
+
+    public void updateToken(TokenUpdateTransactionBody op, long now) {
```

Comment:
Why isn't this method being used in `TokenUpdateTransitionLogic` ? The `transitionFor` method in `TokenUpdateTransitionLogic` should be refactored to use this method

Reply:
The approach of separate logic for HTSPrecompileUpdate was taken mainly because of the TransactionContext and HederaLedger fields, as you can see in TokenUpdateTransitionLogic all failed checks need to adjust the state whereas in the precompile failed checks just revert.

Reply:
I see. But most of the logic seems to be the same and duplicated in two places. It will be nice to reuse the methods from `TokenUpdateLogic.java` in `TokenUpdateTransitionLogic` for future maintenance.

Reply:
It will for sure be better to reuse the UpdateLogic, however, at the moment I can't see a straightforward way to do so, it will be a lot of effort and need to be planned and discussed further. So may be we can do it in a separate PR dedicated for token update refactoring?